File: my_problems/day47-dsa-binarysearch-II/q4_AthMagicalNumber.java
public class q4_AthMagicalNumber {

    // custom function to calculate the GCD
    public static long gcd(long a, long b) {

        // make sure to maintain the value of a < b
        if (a > b) {
            long temp = a;
            a = b;
            b = temp;
        }
        // repeat while a becomes 0
        while (a > 0) {
            // make to maintain a < b
            if (a > b) {
                long temp = a;
                a = b;
                b = temp;
            }
            // update value of a and b
            long temp = a;
            a = b % a;
            b = temp;
        }
        // return final answer
        return b;
    }

    public static long getCountOfMultiplesTillN(long N, long A, long B) {
        long x = N;
        long a = A;
        long b = B;
        // find Ath number multiple of either B or C
        long mult = a * b;
        long lcm = mult / (long) gcd(A, B);

        long ans = (x / a) + (x / b) - (x / lcm);
        return ans;
    }

    public static long AthMagicalNumber(long N, long A, long B) {

        // define search space
        long l = Math.min(A, B);
        long r = l * N;

        long ans = 0;
        while (l <= r) {
            long mid = l + (r - l) / 2;
            long count = getCountOfMultiplesTillN(mid, A, B);
            if (count == N) {
                // update answer and check for better answer on left
                ans = mid;
                r = mid - 1;
            } else if (count < N) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }

        System.out.println(ans);
        return ans;
    }

    public static void main(String[] args) {

        // AthMagicalNumber(10, 2, 3);
        // AthMagicalNumber(2, 2, 3);
        // AthMagicalNumber(4, 2, 3);
        // AthMagicalNumber(6, 5, 7);
        AthMagicalNumber(807414236, 3788, 38141); // expected result 238134151

    }

}

File: my_problems/day47-dsa-binarysearch-II/q4_AthMagicalNumber.js
/**
 * Ath Magical number
 */

const solve = (A, B, C) => {
    let lcmValue = lcm(B, C);
    let start = Math.min(B, C);
    let end = start * A;
    let ans = start;
    let m = 1000000007;

    while (start <= end) {
        let mid = start + Math.floor((end - start) / 2);
        let count = getCount(mid, B, C, lcmValue);
        if (count === A) {
            ans = mid % m;
            end = mid - 1;
        } else if (count > A) {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }
    return ans;
}

const getCount = (n, a, b, lcmValue) => {
    return Math.floor(n / a) + Math.floor(n / b) - Math.floor(n / lcmValue);
}

const lcm = (a, b) => {
    let multiplication = a * b;
    return multiplication / gcd(a, b);
}

const gcd = (a, b) => {

    // make sure a is always min out of 2
    if (a > b) {
        let temp = a;
        a = b;
        b = temp;
    }

    if (a === 0) {
        return b;
    }
    if (a === 1) {
        return 1;
    }

    return gcd(a, b % a);
}

console.log(solve(12, 10, 13));
console.log(solve(19, 11, 13));
console.log(solve(807414236, 3788, 38141));